File: 프로그래머스/3/258705. 산 모양 타일링/산 모양 타일링.java
class Solution {
    public int solution(int n, int[] tops) {
        int answer = 0;
        
        int dp[][] = new int[n+1][2];
        
        dp[0][0] = 1;
        int mod = 10007;
        
        for(int i = 1; i < n+1; i++){
            if(tops[i-1] == 1){
                dp[i][0] = (dp[i-1][0] * 3 + dp[i-1][1] * 2) % mod;
            }else{
                dp[i][0] = (dp[i-1][0] * 2 + dp[i-1][1]) % mod;
            }
            
            dp[i][1] = (dp[i-1][1] + dp[i-1][0]) % mod;
        }
        
        return (dp[n][0] + dp[n][1]) % mod;
    }
}

File: 프로그래머스/3/258705. 산 모양 타일링/산 모양 타일링.js
function solution(n, tops) {
    const mod = 10007;
    
    let dp = new Array({length : n+1}).fill(0);
    let dp3 = new Array({length : n + 1}).fill(0);
    
    dp[1] = tops[0] ? 3 : 2;
    dp3[1] = 1;
    
    for(let i = 2; i < n + 1; i++){
        if(tops[i - 1]){
            dp[i] = (dp[i - 1] * 3 + dp3[i - 1] * 2) % mod;
        }else{
            dp[i] = (dp[i - 1] * 2 + dp3[i - 1]) % mod;
        }
        dp3[i] = (dp[i - 1] % mod) + (dp3[i - 1] % mod);
    }
    return (dp[n] + dp3[n]) % mod;
}